provers/risc0/agent/src/boundless.rs

```diff
@@ -439,163 +584,738 @@ impl Risc0BoundlessProver {
+        // Get from SQLite storage for most up-to-date data
+        match self.storage.list_active_requests().await {
+            Ok(requests) => requests,
+            Err(e) => {
+                tracing::warn!("Failed to get requests from storage, falling back to memory: {}", e);
+                let requests_guard = self.active_requests.read().await;
+                requests_guard.values().cloned().collect()
+            }
+        }
+    }
+
+    /// Get database statistics for monitoring
+    pub async fn get_database_stats(&self) -> AgentResult<crate::storage::DatabaseStats> {
+        self.storage.get_stats().await
+    }
+
+    /// Delete all requests from the database
+    /// Returns the number of deleted requests
+    pub async fn delete_all_requests(&self) -> AgentResult<usize> {
+        let deleted_count = self.storage.delete_all_requests().await?;
+        
+        // Clear in-memory active requests as well
+        self.active_requests.write().await.clear();
+        
+        tracing::info!("Deleted {} requests from database and cleared memory cache", deleted_count);
+        Ok(deleted_count)
+    }
+
+
+    async fn update_impl(&self, elf: Vec<u8>, elf_type: ElfType) -> AgentResult<Vec<u8>> {
+        tracing::info!("Updating ELF for type: {:?}", elf_type);
+        
+        // Basic ELF validation
+        self.validate_elf(&elf)?;
+        
+        let boundless_client = self.create_boundless_client().await?;
+        
+        // Upload the new ELF to storage provider
+        let new_image_url = boundless_client
+            .upload_program(&elf)
+            .await
+            .map_err(|e| AgentError::ProgramUploadError(format!("Failed to upload new ELF: {e}")))?;
+        
+        tracing::info!("Successfully uploaded new ELF to: {}", new_image_url);
+        
+        // Update the appropriate image URL based on ELF type
+        self.update_image_url(elf_type.clone(), new_image_url.clone()).await?;
+        
         let response = Risc0Response {
-            seal: seal.to_vec(),
-            journal: journal.to_vec(),
-            receipt: serde_json::to_string(&boundless_receipt).ok(),
+            seal: vec![], // No seal for update operation
+            journal: new_image_url.as_str().as_bytes().to_vec(), // Return the new URL in journal
+            receipt: None,
         };
-
+        
         let proof_bytes = bincode::serialize(&response).map_err(|e| {
-            AgentError::ResponseEncodeError(format!("Failed to encode response: {e}"))
+            AgentError::ResponseEncodeError(format!("Failed to encode update response: {e}"))
         })?;
+        
         Ok(proof_bytes)
     }
 
-    pub async fn update(&self, _elf: Vec<u8>, _elf_type: ElfType) -> AgentResult<Vec<u8>> {
-        // update elf & upload to storage provider, then update the image_url
-        todo!()
+    /// Basic ELF validation
+    fn validate_elf(&self, elf: &[u8]) -> AgentResult<()> {
```

Comment:
why do we need validate elf?? if elf is generated during compiling?
checking image id makes more sense